cincinnati/src/plugins/internal/graph_builder/openshift_secondary_metadata_parser/plugin.rs

```diff
@@ -148,41 +154,103 @@ impl PluginSettings for OpenshiftSecondaryMetadataParserSettings {
     }
 }
 
+#[derive(Debug, Fail, strum_macros::EnumDiscriminants)]
+#[strum(serialize_all = "snake_case")]
+#[strum_discriminants(derive(Hash, Serialize, Deserialize), serde(rename_all = "snake_case"))]
+pub enum DeserializeDirectoryFilesError {
+    #[error("Error reading {0:?}: {1:#?}")]
+    File(PathBuf, std::io::Error),
+
+    #[error("{0:?}: has an invalid extension {1:?}")]
+    InvalidExtension(PathBuf, String),
+
+    #[error("{0:?}: is missing an extension")]
+    MissingExtension(PathBuf),
+
+    #[error("Failed to deserialize {0:?}: {1:#?}")]
+    Deserialize(PathBuf, serde_yaml::Error),
+}
+
 pub async fn deserialize_directory_files<T>(
     path: &PathBuf,
     extension_re: regex::Regex,
+    disallowed_errors: &HashSet<DeserializeDirectoryFilesErrorDiscriminants>,
 ) -> Fallible<Vec<T>>
 where
     T: DeserializeOwned,
 {
+    use std::sync::Arc;
+    use std::sync::Mutex;
     use tokio::stream::Stream;
     use tokio::stream::StreamExt;
 
+    // Even though we don't use concurrent threads, the usage of async forces us
+    // to guarantee that the error container is thread-safe
+    let error: Arc<Mutex<Option<Error>>> = Arc::new(Mutex::new(None));
```

Comment:
I used [appendlist](https://docs.rs/appendlist/1.4.0/appendlist/) in cincinnati-graph-data PR - perhaps its would be useful here?

Reply:
I'm happy to discuss that approach as well.

I think I want to start backwards, focusing on what Error type do we want in the Result of the function? I see the following options:

(a) A single Error -- this is what I chose
(b) A collection of Error

Choosing (a) it makes sense to me to use [anyhow::Context](https://docs.rs/anyhow/1.0.32/anyhow/trait.Context.html) to append all the errors into a single instance of `anyhow::Error`. This has the advantage of not requiring to iterate through the list of errors later to consolidate them into a single Error.

If we would choose (b) I think using the _appendlist_ crate would make sense as we could likely transform it into a `Vec`.

So the question we're dealing with now: what's better for the consumer of the function?

Reply:
Returning a single error seems to be the obvious choice - we don't need special treatment for the `Vec` of errors, other functions are returning error message as a string.

Its possible that `cincinnati-graph-data` PR will make a change to both "blocked-edges" and "channels" with invalid yamls in both, but it has been extremely rare. As a result I'm fine with throwing a string error and exiting early

Reply:
Also `appendlist` is great for appending without mutation, but I'm not sure if its thread-safe. it cannot be easily converted to `Vec` though, so +1 to current implementation